acceptance: run under bazel

rail opened the pull request:
Add support for running the acceptance tests in Bazel.

Release note: None

rail committed:
acceptance: run under bazel

Add support for running the acceptance tests in Bazel.

Because the acceptance tests use docker to run the tests and use bind
volumes to share certificates and testdata, we run the sets without
using the bazel sandbox.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68838)
<!-- Reviewable:end -->

rickystewart commented:
Looks like you need to `./dev generate bazel` fyi. :)

rickystewart commented on build/bazelutil/acceptance.sh:
```diff
@@ -0,0 +1,42 @@
+  exit 1
+fi
+
+# cd "$BUILD_WORKSPACE_DIRECTORY/$PACKAGE"
```
If you're going to comment out this line, might as well delete it and also the `if [ -z "${BUILD_WORKSPACE_DIRECTORY-}" ];` stanza above.

rail replied:
👍

rickystewart commented on build/teamcity/cockroach/ci/tests/acceptance.sh:
```diff
@@ -0,0 +1,21 @@
+tc_prepare
+set -x
+
+tc_start_block "Run acceptance"
```
Is this tested? I don't see an acceptance build config in TC.

rail replied:
It's "under construction" [here](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_RailBazelAcceptance?branch=68838&mode=builds) :)

rickystewart commented on build/teamcity/cockroach/ci/tests/acceptance.sh:
```diff
@@ -0,0 +1,21 @@
+
+tc_start_block "Run acceptance"
+run_bazel build/teamcity/cockroach/ci/tests/acceptance_impl.sh
+# bazel build //pkg/cmd/bazci --config=ci
```
delete commented-out code

rail replied:
👍

rickystewart commented on build/teamcity/cockroach/ci/tests/acceptance_impl.sh:
```diff
@@ -0,0 +1,11 @@
+set -xeuo pipefail
+
+bazel build //pkg/cmd/bazci //pkg/acceptance:acceptance_test //pkg/cmd/cockroach-short:cockroach-short --config=ci
+find $(bazel info bazel-bin)/.. -ls
```
What's this for? Just for debugguing?

rail replied:
yup, printf debugging :)

rickystewart commented on build/teamcity/cockroach/ci/tests/acceptance_impl.sh:
```diff
@@ -0,0 +1,11 @@
+
+set -xeuo pipefail
+
+bazel build //pkg/cmd/bazci //pkg/acceptance:acceptance_test //pkg/cmd/cockroach-short:cockroach-short --config=ci
```
Also add `--config=crosslinux` -- pretty sure you'll want it because we're shipping that `cockroach-short` to a Docker container

rail replied:
Doh! True story!

rickystewart commented on build/teamcity/cockroach/ci/tests/acceptance_impl.sh:
```diff
@@ -0,0 +1,11 @@
+find $(bazel info bazel-bin)/.. -ls
+cp \
+    $(bazel info bazel-bin)/pkg/acceptance/acceptance_test_/acceptance_test \
+    $(bazel info bazel-bin)/pkg/cmd/cockroach-short/cockroach-short_/cockroach-short \
```
A couple nits:

1. When doing `bazel info bazel-bin`, add `--config=crosslinux`
2. Only do `bazel info bazel-bin --config=crosslinux` once, save it in a variable, and reference that instead of calling into it multiple times. `bazel info` can take a second or so.

rickystewart commented on pkg/acceptance/BUILD.bazel:
```diff
@@ -27,16 +32,18 @@ go_library(
+    data = glob(["testdata/**"]) + ["//pkg/cmd/cockroach-short:cockroach-short"],
     embed = [":acceptance"],
-    tags = ["broken_in_bazel"],
+    # tags = ["broken_in_bazel"],
```
😠

rail replied:
I know! :)

rickystewart commented on pkg/acceptance/BUILD.bazel:
```diff
@@ -27,16 +32,18 @@ go_library(
     ],
-    data = glob(["testdata/**"]),
+    gotags = ["acceptance"],
+    data = glob(["testdata/**"]) + ["//pkg/cmd/cockroach-short:cockroach-short"],
```
Seems wrong, you're building `cockroach-short` twice (once manually, and then once implicitly here).

rail replied:
IIRC, the main idea here was to have proper dependencies set. The manual built should be almost instantaneous?

rickystewart commented on build/bazelutil/acceptance.sh:
```diff
@@ -0,0 +1,42 @@
+# cd "$BUILD_WORKSPACE_DIRECTORY/$PACKAGE"
+
+TEST_WORKSPACE=cockroach PATH="$(dirname $go_bin):$PATH" GOROOT="$(dirname $(dirname $go_bin))" \
+    artifacts/acceptance_test -b artifacts/cockroach-short "$@"
```
Okay, so a requirement for running this test is that we've built the acceptance test binary and the `cockroach-short` binary and that they're sitting in the `artifacts` directory before you run this test. That's unfortunate and not very "Bazel-like". Can you talk about why you structured it this way/maybe add a comment?

If it must be done this way, least you can do is add some tests in this script to make sure the binaries are there and throw a comprehensible error if not.

rail replied:
To be honest I'm a bit lost how to properly bazelize the run part. Ideally we don't want to rerun the tests unless the deps are changed.
I took this approach just to have something that can be run, so I can iterate and see what else is broken. I'd be happy to chat about the potential options for this step.

rickystewart commented on pkg/acceptance/BUILD.bazel:
```diff
@@ -4,7 +4,7 @@ go_library(
     name = "acceptance",
     srcs = [
         "flags.go",
-        "test_main.go",
```
Pretty confused by the changes in this file. Is this PR missing some files or something? Otherwise these changes should not be necessary.

rail commented on build/bazelutil/BUILD.bazel:
```diff
@@ -13,3 +13,21 @@ go_library(
+        "--mount-cockroach-binary",
+    ],
+    data = [
+        # "//pkg/acceptance:acceptance_test",
```
Will do for sure. It's been in a try-fail-try-again mode for now. :)

rail commented on build/bazelutil/acceptance.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Wrap rlocation such that we immediately fail if a dep is not found.
+rlocation_ck () {
+    loc="$(rlocation $1)"
```
Bad copy paste I guess :) Will fix it.

rail commented:
> Looks like you need to `./dev generate bazel` fyi. :)

Interesting. It reformatted my changes (good!) and reverted some - the ones I added to satisfy the `acceptance` go build tag. Hmm...

rail commented:
Duplicate of #71770
